native/src/main/java/io/ballerina/lib/avro/Avro.java

```diff
@@ -0,0 +1,112 @@
+import com.fasterxml.jackson.dataformat.avro.AvroMapper;
+import com.fasterxml.jackson.dataformat.avro.AvroSchema;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.utils.JsonUtils;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.utils.ValueUtils;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BHandle;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.api.values.BTypedesc;
+import org.apache.avro.Schema;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static io.ballerina.lib.avro.ModuleUtils.AVRO_SCHEMA;
+
+public class Avro {
+
+    private Schema nativeSchema;
+
+    public Avro(BString schema) {
+        Schema.Parser parser = new Schema.Parser();
+        this.nativeSchema = parser.parse(schema.getValue());
+    }
+
+    protected Schema getSchema() {
+        return nativeSchema;
+    }
+
+    protected void setSchema(Schema nativeSchema) {
+        this.nativeSchema = nativeSchema;
+    }
+
+    public static Object toAvro(BObject avroObject, Object data) {
+        BHandle handle = (BHandle) avroObject.get(StringUtils.fromString(AVRO_SCHEMA));
+        Avro avro = (Avro) handle.getValue();
+        Schema schema = avro.getSchema();
+        ObjectMapper objectMapper = new ObjectMapper();
+        Object jsonObject;
+        try {
+            jsonObject = generateJsonObject(data, schema, objectMapper);
+            if (jsonObject == null) {
+                return ValueCreator.createArrayValue(new byte[]{0, 0});
+            } else if (Objects.equals(schema.getType(), Schema.Type.FIXED)) {
+                return ValueCreator.createArrayValue(((BArray) jsonObject).getByteArray());
+            }
+            byte[] avroBytes = (new AvroMapper()).writer(new AvroSchema(schema)).writeValueAsBytes(jsonObject);
+            return ValueCreator.createArrayValue(avroBytes);
+        } catch (JsonProcessingException e) {
+            return Utils.createError(e.getMessage());
+        }
+    }
+
+    public static Object fromAvro(BObject avroObject, BArray payload, BTypedesc typeParam) {
+        AvroMapper mapper = new AvroMapper();
+        BHandle handle = (BHandle) avroObject.get(StringUtils.fromString(AVRO_SCHEMA));
+        Avro avro = (Avro) handle.getValue();
+        Schema schema = avro.getSchema();
+        byte[] avroBytes = payload.getByteArray();
+        if (Objects.equals(schema.getType(), Schema.Type.FIXED)) {
+            return ValueUtils.convert(ValueCreator.createArrayValue(avroBytes), typeParam.getDescribingType());
+        }
+        JsonNode deserializedJsonString;
+        try {
+            deserializedJsonString = mapper.readerFor(Object.class).with(new AvroSchema(schema)).readTree(avroBytes);
+        } catch (IOException e) {
+            return Utils.createError(e.getMessage());
```

Comment:
Instead of just forgetting the java exception can't we provide it as the cause for the Ballerina error. Following code segment [1] is a sample to do that.

[1] - [https://github.com/xlibb/module-asb/blob/7490f5327677b5df379597156526e43a1f91b1b2/native/src/main/java/io/xlibb/asb/CommonUtils.java#L42](https://github.com/xlibb/module-asb/blob/7490f5327677b5df379597156526e43a1f91b1b2/native/src/main/java/io/xlibb/asb/CommonUtils.java#L42)